src/mcelog.c

```diff
@@ -0,0 +1,431 @@
+                                              mr.uncorrected_err_total) < 0) {
+    ERROR("%s: add corrected errors meta data failed", MCELOG_PLUGIN);
+    return (-1);
+  }
+  if (plugin_notification_meta_add_signed_int(
+          n, "uncorrected memory timed errors", mr.uncorrected_err_timed) < 0) {
+    ERROR("%s: add corrected timed errors meta data failed", MCELOG_PLUGIN);
+    return (-1);
+  }
+  if (plugin_notification_meta_add_string(n, "uncorrected errors time period",
+                                          mr.uncorrected_err_timed_period) <
+      0) {
+    ERROR("%s: add corrected errors period meta data failed", MCELOG_PLUGIN);
+    return (-1);
+  }
+
+  return (0);
+}
+
+static int mcelog_submit(mcelog_memory_rec_t mr) {
+
+  value_list_t vl = VALUE_LIST_INIT;
+  char type_instance[DATA_MAX_NAME_LEN] = {0};
+  char plugin_instance[DATA_MAX_NAME_LEN] = {0};
+  vl.values_len = 1;
+  vl.time = cdtime();
+
+  sstrncpy(vl.plugin, MCELOG_PLUGIN, sizeof(vl.plugin));
+  sstrncpy(vl.type, "errors", sizeof(vl.type));
+  if (strlen(mr.dimm_name) > 0) {
+    snprintf(plugin_instance, DATA_MAX_NAME_LEN, "%s_%s", mr.location,
+             mr.dimm_name);
+    sstrncpy(vl.plugin_instance, plugin_instance, sizeof(vl.plugin_instance));
+  } else
+    sstrncpy(vl.plugin_instance, mr.location, sizeof(vl.plugin_instance));
+
+  sstrncpy(vl.type_instance, "corrected_memory_errors",
+           sizeof(vl.type_instance));
+  vl.values = &(value_t){.derive = (derive_t)mr.corrected_err_total};
+  plugin_dispatch_values(&vl);
+
+  snprintf(type_instance, DATA_MAX_NAME_LEN, "corrected_memory_errors_in_%s",
+           mr.corrected_err_timed_period);
+  sstrncpy(vl.type_instance, type_instance, sizeof(vl.type_instance));
+  vl.values = &(value_t){.derive = (derive_t)mr.corrected_err_total};
+  plugin_dispatch_values(&vl);
+
+  sstrncpy(vl.type_instance, "uncorrected_memory_errors",
+           sizeof(vl.type_instance));
+  vl.values = &(value_t){.derive = (derive_t)mr.uncorrected_err_total};
+  plugin_dispatch_values(&vl);
+
+  snprintf(type_instance, DATA_MAX_NAME_LEN, "uncorrected_memory_errors_in_%s",
+           mr.uncorrected_err_timed_period);
+  sstrncpy(vl.type_instance, type_instance, sizeof(vl.type_instance));
+  vl.values = &(value_t){.derive = (derive_t)mr.uncorrected_err_total};
+  plugin_dispatch_values(&vl);
+
+  return 0;
+}
+
+static void *poll_worker(__attribute__((unused)) void *arg) {
+  char buf[DATA_MAX_NAME_LEN] = {0};
+  mcelog_memory_rec_t memory_record;
+  _Bool received_data = 0;
+  FILE *p_file;
+
+  struct pollfd poll_fd = {
+      .fd = g_mcelog_config.sock_fd, .events = POLLIN | POLLPRI,
+  };
+
+  while (1) {
+
+    pthread_mutex_lock(&mcelog_thread_lock);
+    if (mcelog_thread_shutdown == 1) {
+      pthread_mutex_unlock(&mcelog_thread_lock);
+      pthread_exit(NULL);
+    }
+    pthread_mutex_unlock(&mcelog_thread_lock);
+    int poll_ret = poll(&poll_fd, 1, POLL_TIMEOUT);
```

Comment:
You mean adding signal handler to thread function  poll_worker ? Or just terminating thread asynchronously sending SIGTERM ?  I have noticed that some plugins use that approach but  I think this is not applicable here since we have to do some cleanup like calling fclose and so on.